### DISCUSSION TITLE
Supporting reference access/refresh tokens

### DISCUSSION THREAD

**Discussion by tnorimat** [keycloak/keycloak#19649] [Ideas]
There are two types of tokens:
  - Self-contained token: including information. Not opaque.
  - Reference token: not including information. Opaque. An authorization server provides information in return for the token. 
 
Currently, Keycloak only supports a self-contained token. How about supporting also reference token?

Motivation for a reference token is as follows:
  - Privacy: Some use cases require not including sensitive data like Personal Identifiable Information (PII) in an access token.
  - Confidentiality: Not disclosing information which a resource server only need to know to 3rd party entities like clients.
  - Security requirement: Some use cases requires a reference token as their requirement.
 
Other solution to resolve the points above is as follows:
  - Lightweight access token: discussed but not yet realized.
    - Discussion: https://github.com/keycloak/keycloak/discussions/9713
    - PR (Draft) : https://github.com/keycloak/keycloak/pull/8914
    - PR (Draft) : https://github.com/keycloak/keycloak/pull/22148
  - Access token encryption: tried but not realized.
    - PR (Closed) : https://github.com/keycloak/keycloak/pull/6796, https://github.com/keycloak/keycloak/pull/7340, https://github.com/keycloak/keycloak/pull/7341

Namely, there are three types of solution: light weight token, token encryption, and reference token.

The basic concept of supporting a reference token is the following:
  - Keycloak continues treating an access and refresh token as a self-contained token internally.
  - On sending a token to a client, Keycloak converts a self-contained token to a reference token.
  - On receiving a token from a client and resource server, Keycloak converts a reference token to a self-contained token.
 
The major concern for supporting a reference token is the following (just mentioned in [Light weight access token's discussion](https://github.com/keycloak/keycloak/discussions/9713))
 - Cache: requires a separate cache to be able to exchange a reference for the token.

To resolve the concern, the following way can be taken:
 - Cache: Keycloak does not implement such the cache by itself. Instead of it, Keycloak defines I/F with the cache and a user of Keycloak prepares it implementing the I/F, which burdens a user of Keycloak who want to use a reference token. However, Keycloak itself and other user who uses a self-contained token as usual are not affected by supporting a reference token.

IMO, to support Light weight access token support and token encryption, we need to modify Keycloak's body codes. I have an idea to support a reference token without modifying keycloak's body codes heavily:
 - Non-invasive way: a reference token is realized by client policies, which means that Keycloak's body codes and other user who uses a self-contained token as usual are not affected by supporting a reference token.

Other points:
 - Flexibility: by using client policies, it is easy to specify which clients use a reference token. Namely, we can select using a reference token or a self-contained token per client.
 - Flexibility: a user of Keycloak can implement their own I/F with the cache as a SPI provider.

I considered specifications of supporting reference access/refresh tokens and wrote them in [the issue](https://github.com/keycloak/keycloak/issues/19650).
I have already implemented PoC codes for it the by following the specifications and confirmed that they work well as I expected.

**Comment by thomasdarimont** [comment]
The old issue [KEYCLOAK-8278](https://issues.redhat.com/browse/KEYCLOAK-8278) also contains some discussions around reference token support for Keycloak.

**Comment by tnorimat** [comment]
@thomasdarimont Thank you Thomas. In my plan, a user can customize the format of a reference token by themselves. For example, a random string, including some information (timestamp, etc...).

**Reply by thomasdarimont** [reply]
Sounds great! I wonder if the default implementation could already provide "signed" and "timestamped" reference tokens that could at least partially validated by a resource server (integrity checks, not timed-out, etc.), even if the rest of the payload is really just a reference to a "session" on the auth server.

**Reply by tnorimat** [reply]
@thomasdarimont In the [PR](https://github.com/keycloak/keycloak/pull/19824), `ReferenceTypeTokenExecutor`:
```
    @Override
    protected String createReferenceTypeAccessToken(TokenManager.AccessTokenResponseBuilder builder) {
        return builder.getAccessToken().getId();
    }

    @Override
    protected String createReferenceTypeRefreshToken(TokenManager.AccessTokenResponseBuilder builder) {
        return builder.getRefreshToken().getId();
    }
```

It seems that what you want to do can be achieved  by copying a builder, modify the copied builder (e,g. drop claims of an access token other than iat, ext), build the copied builder, and return the generated token as a reference token.

**Comment by jbman** [comment]
Concept for reference tokens with own cache SPI sounds good! 
I would like to add, that the client should have a config option to enable the use of reference token. Client policy could just enforce this option, not controlling behavior.

**Reply by tnorimat** [reply]
@jbman It seems to be difficult to implement logic for a reference token on keycloak's body codes. So in the [PR](https://github.com/keycloak/keycloak/pull/19824), I implemented the logic in some executor `ReferenceTypeTokenExecutor` and `AbstractReferenceTypeTokenExecutor`. You can create your own type of a reference token by extending these classes (e.g., use Redis as a external token store instead of a token store whose interface is REST API, a reference token that include minimum information, which might be called "lightweight token")

**Comment by mposolda** [comment]
We discussed the topic of reference tokens with @stianst and @tnorimat.

The big challenge of this reference tokens proposal is how to store the tokens on Keycloak side? The possible options are:
1) Some storage on Keycloak side (EG. infinispan cache or database) - this means lots of data stored on Keycloak side...
2) External storage server, which can communicate with Keycloak by some proprietary HTTP protocol. This is done in current PR by Takashi [1]. Looks like a workaround and will require people to provide their own HTTP server (Not very aligned of the idea of Keycloak to allow most of the capabilities out-of-the-box without requirement of people doing their own development...)
3) Don't store anything on Keycloak, but still allow Introspection endpoint and UserInfo endpoint to return all the needed info based on the access token

The (3) is most ideal solution. This means that access-token (and possibly refresh-token) would be still signed JWT, however it will be very lightweight and may not contain any user private info (PII). Some info below on what may be required to achieve this in Keycloak:

## More flexibility for Introspection endpoint

Current Token Introspection endpoint (and especially `AccessTokenIntrospectionProvider`) doesn't have much flexibility on what claims to return. It currently basically just copies the claims from access-token and returns them in the response. We may need some additional flexibility here and make sure that introspection endpoint can return even claims, which are not present in the access token. We can add flag `Add to Token Introspection` available on most of ProtocolMappers. This means possibly also introducing interface like `TokenIntrospectionTokenMapper` (similar to EG. `UserInfoTokenMapper`, `OIDCIDTokenMapper` and `OIDCAcessTokenMapper`). The flag can be ON by default. Claims would be added by TokenIntrospection just when enabled by particular mapper.

## Possibility to hide sub claim from access token

AFAIK most of the user PII claims are added to access token (and IDToken and UserInfo) by protocol mappers, however the `sub` claim is still added automatically - https://github.com/keycloak/keycloak/blob/21.1.1/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L881 . It can be good if this claim is also added by protocol mapper - it can be perhaps some protocol mapper defined on some realm default client scope (We may need to add new realm default client scope for this).

I am not 100% sure if this capability is needed for reference tokens, but might be useful as it will allow to hide `sub` from access token and/or IDToken, but still have it available on UserInfo or Token Introspection response. Note that `sub` claim does not need on the token as long as there is `session_state` on the token (As UserInfo/TokenIntrospection endpoint can lookup userSession based on `session_state` and then lookup user from it).


## Client policy executor

The Takashi's PR [1] already uses client policies for reference tokens. We can perhaps have an executor, which will allow to:
- Automatically disable all the switches `Add to Access Token` for the current session. This may be needed because protocol mappers are defined on client scopes like `profile`, which are shared for all the realm clients - the clients, which want lightweight/reference access tokens without PII as well as clients, which want normal access tokens with PII. This means that even if protocol mappers on `profile` client scope (and other client scopes) are still ON for `Add to Access Token`, this executor will disable it just for this request of this client and hence will cause claims not to be added to accessToken (and refresh token).

- IMO we can still keep abstract executor like in PR [1] and then it's subclass, which will do the filtering of access token as described in previous point. This may allow more flexibility if someone still preferres access token in their own format. However IMO I don't think that we need to implement token lookups from external HTTP server as currently done in the PR [1].

## Milestones

We can possibly implement this in milestones, so it is not 1 big PR contains everything. I can imagine something like:

Milestone 1: Introspection endpoint

Milestone 2: Client policy executor (maybe just update of the PR [1] is OK), which will switch `Add to Access Token` OFF for this request and will manually remove `sub` claim

Milestone 3: Protocol mapper (and client scope) for `sub` (Optional, not 100% sure if this is needed...)


[1] https://github.com/keycloak/keycloak/pull/19824

**Reply by sschu** [reply]
I like your reasoning here @mposolda
Just to make sure I got the idea right: You would still use JWTs and keep everything out that Keycloak stores in the session. The bare minimum to get as close to a reference token as possible would probably be:

- iss
- exp
- aud
- iat
- session_state
- maybe jti

So in a way, this would also implement the light-weight access token idea.
One could try to get size down even further by using some custom format with only the bare minimum amount of data (probably session_state, iat, exp) and encrypt this with a symmetric key. However, this is probably not worth it as I am not sure the result will be smaller after crypto and you could also just create an encrypted jwt to make it opaque.

**Reply by mposolda** [reply]
@sschu Thanks for listing the "bare minimum" of claims, which would be needed for the reference token. It seems to me that your list is accurate.

I also wonder that this "reference token" can be encrypted, but also it could be signed by the realm symmetric key (H256 or so) instead of asymmetric key, as the only party, which needs to understand "reference token" is Keycloak itself. To clients as well as to resource-server, the reference token is opaque string.

I wonder that Keycloak client executor for "reference token" can have option for optionally treat access token as INTERNAL category (see https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java#L190-L208 on what is internal category), which should be sufficient. Also it can possibly have option for encrypt access token (or maybe same option for both cases would work just fine). That is possible another follow-up, which we can add.

**Reply by cgeorgilakis** [reply]
For required claims, I propose to take into account [Data Structure for OAuth 2.0 Access Tokens](https://datatracker.ietf.org/doc/html/rfc9068#name-data-structure)](https://datatracker.ietf.org/doc/html/rfc9068#name-data-structure) at least for already supported claims in Keycloak.

Maybe this is a related to this github discussion but a seperated discussion. Our goal is to be compliant for all Oauth flow with this specification with needed Keycloak configuration.

From docummentation :

> - iss REQUIRED - as defined in [Section 4.1.1](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.1) of [[RFC7519](https://datatracker.ietf.org/doc/html/rfc7519)].
> - exp REQUIRED - as defined in [Section 4.1.4](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.4) of [[RFC7519](https://datatracker.ietf.org/doc/html/rfc7519)].
> - aud REQUIRED - as defined in [Section 4.1.3](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.3) of [[RFC7519](https://datatracker.ietf.org/doc/html/rfc7519)]. See [Section 3](https://datatracker.ietf.org/doc/html/rfc9068#JWTATLRequest) for indications on how an authorization server should determine the value of "aud" depending on the request.
> - sub
REQUIRED - as defined in [Section 4.1.2](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.2) of [[RFC7519](https://datatracker.ietf.org/doc/html/rfc7519)]. In cases of access tokens obtained through grants where a resource owner is involved, such as the authorization code grant, the value of "sub" SHOULD correspond to the subject identifier of the resource owner. In cases of access tokens obtained through grants where no resource owner is involved, such as the client credentials grant, the value of "sub" SHOULD correspond to an identifier the authorization server uses to indicate the client application. See [Section 5](https://datatracker.ietf.org/doc/html/rfc9068#SecurityConsiderations) for more details on this scenario. Also, see [Section 6](https://datatracker.ietf.org/doc/html/rfc9068#PrivacyConsiderations) for a discussion about how different choices in assigning "sub" values can impact privacy.
> -  client_id REQUIRED - as defined in [Section 4.3](https://www.rfc-editor.org/rfc/rfc8693#section-4.3) of [[RFC8693](https://datatracker.ietf.org/doc/html/rfc8693)].
> -  iat REQUIRED - as defined in [Section 4.1.6](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.6) of [[RFC7519](https://datatracker.ietf.org/doc/html/rfc7519)]. This claim identifies the time at which the JWT access token was issued.
> -  jti REQUIRED - as defined in [Section 4.1.7](https://www.rfc-editor.org/rfc/rfc7519#section-4.1.7) of [[RFC7519](https://datatracker.ietf.org/doc/html/rfc7519)].

**Reply by sschu** [reply]
For reference tokens, this whole list is not needed because as @mposolda said, only Keycloak has to understand the token and no external party. So the list only has to contain stuff Keycloak needs to make sure it is a valid token: `iss` as absolute minimum. `exp`,` aud` and `iat` are for proper JWT validation. To convey information Keycloak does not store in the user session, `iat` and `exp` are needed. The rest can be derived from session information using `session_state` (including `aud`).

Still, it should be possible to create a token according to https://datatracker.ietf.org/doc/html/rfc9068#name-data-structure for light-weight or general access token usage if needed.

@mposolda I am not 100% sure from the top of my head if the `client_id` can be derived with client credentials grant in case you want to add this to the introspection response, I remember client sessions got removed in that case some time ago.

**Reply by thomasdarimont** [reply]
Instead of using the [deprecated session_state](https://github.com/keycloak/keycloak/blob/main/core/src/main/java/org/keycloak/representations/IDToken.java#L69) claim I'd recommend to use the `sid` claim instead for referencing the internal keycloak user / client session.

**Reply by sschu** [reply]
If it contains the same value I am totally with you @thomasdarimont.

**Reply by thomasdarimont** [reply]
Yes, they contain the same values
````java
    // session_state is deprecated, sid should be used instead
    @JsonProperty(SESSION_STATE) // "session_state"
    @JsonAlias(SESSION_ID) // "sid"
    protected String sessionState;
````

**Comment by tnorimat** [comment]
@mposolda I agree with your milestone. I will create an issue per each.
As for Milestone 2, it is needed to enhance client policies. [The PR [1]](https://github.com/keycloak/keycloak/pull/19824) includes this enhancement. so I would like to add it to Milestone as Milestone 0. I will create an issue of it. In Milestone 2, I can create an executor based on Milestone 0.

**Comment by cgeorgilakis** [comment]
Token Introspection enchance is also needed in lightway access token (see [realted discussion](https://github.com/keycloak/keycloak/discussions/9713). We believe that Keycloak should also support this. We could make a PR for Token Introspection enhance if we agree in business. We believe that a field 'Add to introspection response' field in most of protocol mappers is enough. This field will be shown only if 'Add to access token' is false. Token Introspection endpoint will be enchanced by mappers of scopes  that 'Add to introspection response' field is true.

Keycloak already support changing sub claim by creating a protocol mapper with sub claim. We use it. Sub claim is required based on [JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens](https://datatracker.ietf.org/doc/html/rfc9068#name-data-structure). So I do not believe that a user must be able to remove it. We only add preferred_username claim in client credentials flow for being able to use this access token in Token Introspection and UserInfo Endpoint. A note in documentation for this maybe is enough.

**Reply by sschu** [reply]
If I got the proposal right, the idea is actually to implement light-weight access tokens instead of pure reference tokens.
I could imagine there are use cases where you would like to have a field in the access token and the introspection response so I would keep both options independent.
I also see use cases where it makes sense to not have a sub claim and for example only testify you are over 18 or something like that. The whole idea of self-sovereign identity gives users full control over what kind of information they want to reveal about themselves so not revealing a unique user identifier is definitely one of the options and should be supported by Keycloak.

**Reply by cgeorgilakis** [reply]
I propose to support both light-weight access tokens and reference tokens. My comment is for making an issue - PR for Token Introspection enchance based on @mposolda comment. Token Introspection enchance is only needed for light-weight access token.

As far as sub claim, the most important is that Keycloak latest version add sub claim in access token and give the possibility to change its value. It is compliant with [JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens](https://datatracker.ietf.org/doc/html/rfc9068#name-data-structure). Supporting removing sub claim as extra configuration, is no problem for me. Only to mention that sub or preferred_username claim is only needed for client credentials flow ( service account client).

**Reply by sschu** [reply]
But you need token introspection also for reference tokens or how would you like to find out what kind of access the token grants?

**Reply by mposolda** [reply]
@cgeorgilakis @sschu Yes, I think that token introspection would be needed also for reference tokens if we implement it in a way described above (which kind of assume that reference token is JWT to avoid the need of storing additional data on Keycloak side).

I see reference-token as opaque string from the perspective of client/resource-server. Hence it does not need to be valid JWT in theory and hence missing `sub` claim should be fine. On the other hand, the missing `sub` claim might be also "optional" string as for "lightweight access token", we may probably want to support the case when "lightweight access token" is valid JWT.

**Reply by sschu** [reply]
Exactly, all we need to support is removing the sub claim on the mappers and/or via client policy.
